// @chainlink/belt/lib/src/services/compilers/ethers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const ts_generator_1 = require("ts-generator");
const TypeChain_1 = require("typechain/dist/TypeChain");
const utils_1 = require("../utils");
/**
 * Generate ethers.js contract abstractions for all of the solidity versions under a specified contract
 * directory.
 *
 * @param conf The application configuration, e.g. where to read artifacts, where to output, etc..
 */
async function compileAll(conf) {
    const cwd = process.cwd();
    return Promise.all(utils_1.getArtifactDirs(conf).map(async ({ dir }) => {
        const c = compiler(conf, cwd, dir);
        await ts_generator_1.tsGenerator({ cwd, loggingLvl: 'verbose' }, c);
    }));
}
exports.compileAll = compileAll;
/**
 * Create a typechain compiler instance that reads in a subdirectory of artifacts e.g. (abi/v0.4, abi/v0.5.. etc)
 * and outputs ethers contract abstractions under the same version prefix, (ethers/v0.4, ethers/v0.4.. etc)
 *
 * @param config The application level config for compilation
 * @param cwd The current working directory during this programs execution
 * @param subDir The subdirectory to use as a namespace when reading artifacts and outputting
 * contract abstractions
 */
function compiler({ artifactsDir, contractAbstractionDir }, cwd, subDir) {
    return new TypeChain_1.TypeChain({
        cwd,
        rawConfig: {
            files: path_1.join(artifactsDir, subDir, '**', '*.json'),
            outDir: path_1.join(contractAbstractionDir, 'ethers', subDir),
            target: 'ethers',
        },
    });
}
