// @chainlink/belt/lib/src/services/compilers/truffle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const shelljs_1 = require("shelljs");
const utils_1 = require("../utils");
/**
 * Generate @truffle/contract abstractions for all of the solidity versions under a specified contract
 * directory.
 *
 * @param conf The application configuration, e.g. where to read artifacts, where to output, etc..
 */
async function compileAll(conf) {
    utils_1.getArtifactDirs(conf).forEach(({ dir }) => {
        getContractPathsPer(conf, dir).forEach(p => {
            const json = utils_1.getJsonFile(p);
            const fileName = path_1.basename(p, '.json');
            const file = fillTemplate(fileName, {
                contractName: json.contractName,
                abi: json.compilerOutput.abi,
                evm: json.compilerOutput.evm,
                metadata: json.compilerOutput.metadata,
            });
            write(path_1.join(conf.contractAbstractionDir, 'truffle', dir), fileName, file);
        });
    });
}
exports.compileAll = compileAll;
/**
 * Create a truffle contract abstraction file
 *
 * @param contractName The name of the contract that will be exported
 * @param contractArgs The arguments to pass to @truffle/contract
 */
function fillTemplate(contractName, contractArgs) {
    return `'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
const contract = require('@truffle/contract')
const ${contractName} = contract(${JSON.stringify(contractArgs, null, 1)})

if (process.env.NODE_ENV === 'test') {
  try {
    eval('${contractName}.setProvider(web3.currentProvider)')
  } catch (e) {}
}

exports.${contractName} = ${contractName}
`;
}
function getContractPathsPer({ artifactsDir }, version) {
    return [...shelljs_1.ls(path_1.join(artifactsDir, version, '/**/*.json'))];
}
function write(outPath, fileName, file) {
    shelljs_1.mkdir('-p', outPath);
    fs_1.writeFileSync(path_1.join(outPath, `${fileName}.js`), file);
}
